crossbeam-utils/src/atomic/atomic_cell.rs

```diff
@@ -691,124 +928,275 @@ macro_rules! impl_arithmetic {
-    /// assert_eq!(a.load(), true);
-    ///
-    /// assert_eq!(a.fetch_nand(true), true);
-    /// assert_eq!(a.load(), false);
-    ///
-    /// assert_eq!(a.fetch_nand(false), false);
-    /// assert_eq!(a.load(), true);
-    /// ```
-    #[inline]
-    pub fn fetch_nand(&self, val: bool) -> bool {
-        let a = unsafe { &*(self.as_ptr() as *const AtomicBool) };
-        a.fetch_nand(val, Ordering::AcqRel)
-    }
+atomic_maybe_uninit::cfg_has_atomic_16! {
+    impl_arithmetic!(u16, atomic::AtomicU16, "let a = AtomicCell::new(7u16);");
+    impl_arithmetic!(i16, atomic::AtomicI16, "let a = AtomicCell::new(7i16);");
+}
+atomic_maybe_uninit::cfg_no_atomic_16! {
+    impl_arithmetic!(u16, fallback, "let a = AtomicCell::new(7u16);");
+    impl_arithmetic!(i16, fallback, "let a = AtomicCell::new(7i16);");
+}
 
-    /// Applies logical "or" to the current value and returns the previous value.
-    ///
-    /// # Examples
-    ///
-    /// ```
-    /// use crossbeam_utils::atomic::AtomicCell;
-    ///
-    /// let a = AtomicCell::new(false);
-    ///
-    /// assert_eq!(a.fetch_or(false), false);
-    /// assert_eq!(a.load(), false);
-    ///
-    /// assert_eq!(a.fetch_or(true), false);
-    /// assert_eq!(a.load(), true);
-    /// ```
-    #[inline]
-    pub fn fetch_or(&self, val: bool) -> bool {
-        let a = unsafe { &*(self.as_ptr() as *const AtomicBool) };
-        a.fetch_or(val, Ordering::AcqRel)
-    }
+atomic_maybe_uninit::cfg_has_atomic_32! {
+    impl_arithmetic!(u32, atomic::AtomicU32, "let a = AtomicCell::new(7u32);");
+    impl_arithmetic!(i32, atomic::AtomicI32, "let a = AtomicCell::new(7i32);");
+}
+atomic_maybe_uninit::cfg_no_atomic_32! {
+    impl_arithmetic!(u32, fallback, "let a = AtomicCell::new(7u32);");
+    impl_arithmetic!(i32, fallback, "let a = AtomicCell::new(7i32);");
+}
 
-    /// Applies logical "xor" to the current value and returns the previous value.
-    ///
-    /// # Examples
-    ///
-    /// ```
-    /// use crossbeam_utils::atomic::AtomicCell;
-    ///
-    /// let a = AtomicCell::new(true);
-    ///
-    /// assert_eq!(a.fetch_xor(false), true);
-    /// assert_eq!(a.load(), true);
-    ///
-    /// assert_eq!(a.fetch_xor(true), true);
-    /// assert_eq!(a.load(), false);
-    /// ```
-    #[inline]
-    pub fn fetch_xor(&self, val: bool) -> bool {
-        let a = unsafe { &*(self.as_ptr() as *const AtomicBool) };
-        a.fetch_xor(val, Ordering::AcqRel)
-    }
+atomic_maybe_uninit::cfg_has_atomic_64! {
+    #[cfg(not(crossbeam_no_atomic_64))]
+    impl_arithmetic!(u64, atomic::AtomicU64, "let a = AtomicCell::new(7u64);");
+    #[cfg(not(crossbeam_no_atomic_64))]
+    impl_arithmetic!(i64, atomic::AtomicI64, "let a = AtomicCell::new(7i64);");
+    #[cfg(crossbeam_no_atomic_64)]
+    impl_arithmetic!(u64, fetch_update, "let a = AtomicCell::new(7u64);");
+    #[cfg(crossbeam_no_atomic_64)]
+    impl_arithmetic!(i64, fetch_update, atomic::AtomicI64, "let a = AtomicCell::new(7i64);");
```

Comment:
With `fetch_update` it doesn't take a type, does it?

Also there's a subtle invariant here that `fallback` is only used if regular loads/stores also use the fallback, right? How is that ensured?